x-pack/plugins/security/server/authentication/authenticator.ts

```diff
@@ -378,40 +385,77 @@ export class Authenticator {
                 providerNameSuggestedByHint
               )}`
             : ''
+        }${
+          existingSession.error instanceof SessionExpiredError
+            ? `&${LOGOUT_REASON_QUERY_STRING_PARAMETER}=${encodeURIComponent(
+                existingSession.error.code
+              )}`
+            : ''
         }`
       );
     }
 
+    const requestIsRedirectable = canRedirectRequest(request);
+
     const suggestedProviderName =
-      existingSessionValue?.provider.name ??
+      existingSession.value?.provider.name ??
       request.url.searchParams.get(AUTH_PROVIDER_HINT_QUERY_STRING_PARAMETER);
     for (const [providerName, provider] of this.providerIterator(suggestedProviderName)) {
       // Check if current session has been set by this provider.
       const ownsSession =
-        existingSessionValue?.provider.name === providerName &&
-        existingSessionValue?.provider.type === provider.type;
+        existingSession.value?.provider.name === providerName &&
+        existingSession.value?.provider.type === provider.type;
 
       const authenticationResult = await provider.authenticate(
         request,
-        ownsSession ? existingSessionValue!.state : null
+        ownsSession ? existingSession.value!.state : null
       );
 
       if (!authenticationResult.notHandled()) {
         const sessionUpdateResult = await this.updateSessionValue(request, {
           provider: { type: provider.type, name: providerName },
           authenticationResult,
-          existingSessionValue,
+          existingSessionValue: existingSession.value,
         });
-        return enrichWithUserProfileId(
-          canRedirectRequest(request)
-            ? this.handlePreAccessRedirects(request, authenticationResult, sessionUpdateResult)
-            : authenticationResult,
-          sessionUpdateResult ? sessionUpdateResult.value : null
-        );
+
+        if (requestIsRedirectable) {
+          if (
+            // TODO: Should this just be `instanceof SessionError`?
+            existingSession.error instanceof SessionExpiredError &&
+            authenticationResult.redirectURL?.startsWith(
+              `${this.options.basePath.get(request)}/login`
+            )
+          ) {
+            authenticationResult.redirectURL += `&${LOGOUT_REASON_QUERY_STRING_PARAMETER}=${encodeURIComponent(
+              existingSession.error.code
+            )}`;
```

Comment:
An alternative to making the `redirectUrl` editable (as you can see in the `AuthenticationResult` class) would be to simply create a new `AuthenticationResult` object with the modified URL based off the old object:

```diff
diff --git a/x-pack/plugins/security/server/authentication/authentication_result.test.ts b/x-pack/plugins/security/server/authentication/authentication_result.test.ts
index 8420ae6f5de..73465a9fdaf 100644
--- a/x-pack/plugins/security/server/authentication/authentication_result.test.ts
+++ b/x-pack/plugins/security/server/authentication/authentication_result.test.ts
@@ -296,14 +296,6 @@ describe('AuthenticationResult', () => {
       expect(authenticationResult.userProfileGrant).toBe(userProfileGrant);
       expect(authenticationResult.error).toBeUndefined();
     });
-
-    it('should be able to update the redirectURL after the AuthenticationResult has been created', () => {
-      const newRedirectURL = '/some/other/url';
-      const authenticationResult = AuthenticationResult.redirectTo('/redirect/url');
-      authenticationResult.redirectURL = newRedirectURL;
-
-      expect(authenticationResult.redirectURL).toBe(newRedirectURL);
-    });
   });
 
   describe('shouldUpdateState', () => {
diff --git a/x-pack/plugins/security/server/authentication/authentication_result.ts b/x-pack/plugins/security/server/authentication/authentication_result.ts
index f8ba295c055..62b93a0f0c3 100644
--- a/x-pack/plugins/security/server/authentication/authentication_result.ts
+++ b/x-pack/plugins/security/server/authentication/authentication_result.ts
@@ -222,13 +222,6 @@ export class AuthenticationResult {
     return this.options.redirectURL;
   }
 
-  /**
-   * URL that should be used to redirect user to complete authentication (only available for `redirected` result).
-   */
-  public set redirectURL(url) {
-    this.options.redirectURL = url;
-  }
-
   /**
    * Constructor is not supposed to be used directly, please use corresponding static factory methods instead.
    * @param status Indicates the status of the authentication result.
diff --git a/x-pack/plugins/security/server/authentication/authenticator.ts b/x-pack/plugins/security/server/authentication/authenticator.ts
index 49db9f05ec3..d6853988fc9 100644
--- a/x-pack/plugins/security/server/authentication/authenticator.ts
+++ b/x-pack/plugins/security/server/authentication/authenticator.ts
@@ -406,7 +406,7 @@ export class Authenticator {
         existingSession.value?.provider.name === providerName &&
         existingSession.value?.provider.type === provider.type;
 
-      const authenticationResult = await provider.authenticate(
+      let authenticationResult = await provider.authenticate(
         request,
         ownsSession ? existingSession.value!.state : null
       );
@@ -426,9 +426,11 @@ export class Authenticator {
               `${this.options.basePath.get(request)}/login`
             )
           ) {
-            authenticationResult.redirectURL += `&${LOGOUT_REASON_QUERY_STRING_PARAMETER}=${encodeURIComponent(
-              existingSession.error.code
-            )}`;
+            authenticationResult = AuthenticationResult.redirectTo(
+              authenticationResult.redirectURL + `&${LOGOUT_REASON_QUERY_STRING_PARAMETER}=${encodeURIComponent(
+                existingSession.error.code
+              )}`
+            );
           }
           return enrichWithUserProfileId(
             this.handlePreAccessRedirects(request, authenticationResult, sessionUpdateResult),
```

Reply:
I think I slightly prefer this approach. One nit here: Should we instead be testing that the redirect url starts with `${this.options.basePath.get(request)}/login?` instead of `${this.options.basePath.get(request)}/login` (note the addition of the `?`).

We are appending a query parameter which assumes that the redirect URL already contains at least one other query parameter.

Reply:
I changed the code to use this approach now and included your check for `?`